packages/block-editor/src/hooks/layout.js

```diff
@@ -153,17 +152,22 @@ function LayoutPanel( { setAttributes, attributes, name: blockName } ) {
 		return null;
 	}
 
+	// Layout block support comes from the block's block.json.
 	const layoutBlockSupport = getBlockSupport(
 		blockName,
 		layoutBlockSupportKey,
 		{}
 	);
+	const blockSupportAndThemeSettings = {
+		...layoutSettings,
+		...layoutBlockSupport,
+	};
 	const {
 		allowSwitching,
-		allowEditing = allowEditingSetting ?? true,
+		allowEditing = true,
 		allowInheriting = true,
 		default: defaultBlockLayout,
-	} = layoutBlockSupport;
+	} = blockSupportAndThemeSettings;
```

Comment:
It looks like this changes the logic slightly in that `layoutBlockSupport` now always overrides `layoutSettings` and so setting `allowEditing` in `theme.json` will no longer override the `block.json` setting?

Is it possible to do something like:

* Grab the block level (block.json) settings
* Iterate over the desired `theme.json` settings and if they're set to false, apply them over the `block.json` settings

That way if `allowCustomContentSize` is explicitly set to `true` in `block.json` but `theme.json` sets it to false, the latter wins. But if they're set the other way around `block.json` wins (essentially, `false` always wins).

Just a quick thought, though, happy to give this PR a more thorough review next week! 🙂

Reply:
Yeah that's what I was referring to in my previous comment! In principle, none of those settings should be explicitly enabled as their defaults are always `true`, and this will work as it is for all core blocks. Of course there's no way to control what third party block providers do with their settings, so e.g. a theme that disables controls at a global level won't be able to override a hypothetical custom block that explicitly enables settings. I don't see that as a huge issue though.

There's also the slight problem of `contentSize` and `wideSize` being set to false for all blocks in the settings. Though it's unlikely that a block would set them explicitly in its block.json, that's not something we'd want to override if it _were_ set.

Reply:
Thanks for clarifying!

Reply:
I was a little confused by the `false` value for `contentSize` and `wideSize`, and I think I might have gone down a rabbithole 😄

From looking where it's set ([here](https://github.com/WordPress/gutenberg/blob/08de5b37180ef5ed5eee16ad8a2a34809744a466/packages/block-editor/src/components/global-styles/hooks.js#L310-L316)), I think I see why that's happening as a default now. It seems the assumption in `useSettingsForBlockElement` might have been that `contentSize` and `wideSize` would work similarly to other settings in that they'd be explicitly opted into. So they default to `false` in that hook, but a block might set either of these to `true` in `block.json`.

I was going to say, what if we used `contentSize` and `wideSize` as the boolean flag to show / hide the controls instead of `allowCustomContentSize`, but I'm not sure that would work because we already use these props at the root level to define the _value_ of these fields instead of their visibility status. It's a bit unfortunate, because looking at it now, I wonder if it'd have made sense to have `contentSize` and `wideSize` as values under `styles` and to keep the boolean status in `settings` so that it'd match how padding/margin works? 

At any rate, keeping a separate `allowCustomContentSize` prop (or `allowContentSize`, whichever naming works) sounds like the pragmatic path forward to me.

In terms of the issue of expanding values and passing them down via the `layoutBlockSupport` prop on `layoutType.inspectorControls`, would it be safer to merge just the values that we want to grab from `settings` for now? So similar to how it previously did `allowEditing = allowEditingSetting ?? true,` we might construct it something like:

```
	const blockSupportAndThemeSettings = {
		...layoutBlockSupport,
		allowEditing: layoutSettings.allowEditing ?? true,
		allowCustomContentSize: layout.settings.allowCustomContentSize ?? true,
	};
```

I haven't tried that out, so just an idea! And apologies if I'm overthinking it, I got quite lost looking up those default `false` values 😅

Reply:
That could work! We'll potentially have to do the same for all the other `allow...` settings. I guess there's not that many of them so it's not a huge issue.

Reply:
If I understand correctly, that's essentially what my previous code was doing; might as well revert the commit. I'd like to avoid iterating over specific settings if possible.

Reply:
> If I understand correctly, that's essentially what my previous code was doing; might as well revert the commit.

It's pretty close, but from re-testing, two things that aren't covered in the previous state are:

* If a value is explictly set to `true` in `block.json`, it can't be overridden by `theme.json`
* If we wish to exclude other values in `layoutSettings` that we don't want to pass down (like the `false` values for `contentSize` and `wideSize`)

Would it be better for consistency with other blocks supports to allow `theme.json` to override `true` values, where for example if `spacing.padding` is set to `true` in `block.json`, then it can still be overridden by `theme.json`?

> I'd like to avoid iterating over specific settings if possible.

How come? I was thinking it could be good to be specific about the settings so that we know which ones are handled, otherwise it might be unexpected if or when someone goes to add additional properties that might expect different behaviour 🤔

Reply:
I'd like to avoid the iteration in order to keep the logic clear and ensure existing settings such as `allowJustification` or `allowOrientation` can be enabled just by adding them to `VALID_SETTINGS` (which I was planning to do in a follow-up PR)

Also, ideally the order of things should be predictable and not vary per setting; in this case, it seems best for `block.json` to override `theme.json` settings so that blocks that disable controls aren't overridden. In my mind it would be unexpected if different orders were applied to different settings.

> If a value is explictly set to true in block.json, it can't be overridden by theme.json

Core blocks never explicitly set these values to `true`, and I wouldn't expect most custom blocks to do so either (this can be clarified in documentation to help extenders make good decisions, but regardless, if custom block authors really want controls to always be available then I'd say that's up to them, because they always have the option to create their own controls if core ones don't do what they want)

> If we wish to exclude other values in layoutSettings that we don't want to pass down (like the false values for contentSize and wideSize)

`contentSize` and `wideSize` ending up in this layout settings object shouldn't affect anything as we're not using them at all, unless I'm missing something! The only place this object is being used is in the layout inspector controls.

Reply:
Okay, great, thanks for thinking this through with me! That all sounds reasonable to me, especially once we document it, and it certainly is more readable the way that you've done it 👍

I suppose the last question then is that pesky naming question 😄

Reply:
😅 

I honestly don't have any better ideas, especially given the other layout controls all follow the same pattern. And Lightbox support is [also using](https://github.com/WordPress/gutenberg/blob/trunk/lib/theme.json#L278) `allowEditing` as of 6.4.

Reply:
Me either. I don't feel strongly about the naming, but I'd personally lean toward either `allowCustomContentSize` or `allowContentSize` since we have that prefix for the other layout controls, as you mention, and we likely want parity between what `block.json` uses and what `theme.json` uses. The thing that I like about `allow` as a prefix is that it's clearer it's about what's exposed in the UI, rather than what's output. I think the layout support is possibly a bit unique in this respect, where it can be helpful to have that extra clarity.

Also, the word `custom` in `allowCustomContentSize` is good, because we're still allowing a user to switch between `flow` and `constrained` layout types, where content size is output via the default layout styles... so of all the available options so far, I think `allowCustomContentSize` is my favourite.